### DISCUSSION TITLE
Weird behavior with referential actions

### DISCUSSION THREAD

**Discussion by devskar** [prisma/prisma#18791] [Q&A]
I want to create a user and account model. An account should be able to exist without a user, but a user should always have an account. My models look like this:

```ts
model Account {
  id        String   @id @default(cuid())
  user      User?    @relation(fields: [userId], references: [id], onDelete: SetNull)
  userId    String?  @unique
}

model User {
  id        String  @id @default(cuid())
  username  String  @unique
  Account   Account
}
```
For some reason prisma forces me to make the account field of the User model optional, which i do not want. Also i would like to specify that a user should be deleted when the associated account is deleted, but it seems like i can only set the onDelete value on one side of the relation. How do I solve this?

**Comment by nurul3101** [comment]
Hello @devskar 👋 

Can you try this schema?

```prisma
model Account {
  id     String  @id @default(cuid())
  userId String? @unique
  User   User?
}

model User {
  id       String  @id @default(cuid())
  username String  @unique
  Account  Account @relation(fields: [id], references: [userId], onDelete: Cascade)
}
```

This should satisfy the requirements you mentioned.

**Reply by devskar** [reply]
Hey @nurul3101 thank you this works! I would have some follow up questions:
1. How would the account behave when the user gets deleted?
2. How could I change that behaviour (since I can only set the `@relation` attribute once afaik?
3. Is there a general rule to follow on where to set the `@relation` attribute?

**Comment by devskar** [comment]
Okay this does not seem to work correctly or I have a dumb mistake somewhere. 
When I try to add a user to an account:
```ts
// create user
	const updatedAccount = await prisma.account.update({
		where: {
			id: account.id,
		},
		data: {
			user: {
				create: {
					username,
				},
			},
		},
		include: {
			user: true,
		},
	});
```
I am getting this error: Null constraint violation on the fields: (`id`)

Here is my complete user/account schema as reference:

```ts
model Account {
  id        String   @id @default(cuid())
  sub       String
  provider  Provider
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  userId    String?  @unique
  user      User?

  @@unique([sub, provider])
}

model User {
  id       String  @id @default(cuid())
  username String  @unique
  spots    Spot[]
  account  Account @relation(fields: [id], references: [userId], onDelete: Cascade)
}
```

Also the database is completely empty, except the account the user should be created for. And the id of the account is also correct.

@nurul3101 fyi

**Reply by devskar** [reply]
prisma keeps setting the id of the user to null. Is that the expected behavior?

**Reply by devskar** [reply]
Also the `account` field is still optional in the generate typescript types.

**Reply by devskar** [reply]
I think this is all somehow related to #13885

**Reply by nurul3101** [reply]
Do you get the same error if you create a user and an account in separate queries and then try to connect account to a user?

**Reply by devskar** [reply]
@nurul3101 Okay so: 
```ts
const updatedAccount = await prisma.account.update({
  where: {
	  id: account.id,
  },
  data: {
	  user: {
		  create: {
			  username,
		  },
	  },
  },
  include: {
	  user: true,
  },
});

```

```ts
const user = await prisma.user.create({
  data: {
	  username,
	  account: {
		  connect: {
			  id: account.id,
		  },
	  },
  },
});
```
```ts
const user = await prisma.user.create({ // <- throws here
  data: {
	  username,
  },
});

const updatedAccount = await prisma.account.update({
  where: {
	  id: account.id,
  },
  data: {
	  userId: user.id,
  },
  include: {
	  user: true,
  },
});
```
all result in:

```
Invalid `prisma.account.update()` invocation / Invalid `prisma.user.create()` invocation:
Null constraint violation on the fields: (`id`)
```

**Reply by nurul3101** [reply]
Hey 👋 

Somehow this conversation slipped through the cracks.
Are you still getting this error?

**Reply by devskar** [reply]
@nurul3101 

Hey no worries. Yes and no. I could not figure out a solution so I basically just skipped implementing onDelete and onUpdate settings and I plan to pick it up at another point again. I would still be very much interested in a solution to this problem. I can try to create a minimal Example to this problem later today.

**Reply by nurul3101** [reply]
A minimal example would be great!

**Reply by nurul3101** [reply]
Please feel free to share a minimal example in case you have availability.
